clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -232,6 +235,16 @@ private void onResponse(final ConsumerGroupHeartbeatResponse response, long curr
             this.heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
             this.heartbeatRequestState.resetTimer();
             this.membershipManager.onHeartbeatResponseReceived(response.data());
+            if (previousGroupMetadataUdateEvent == null ||
+                previousGroupMetadataUdateEvent.memberEpoch() != membershipManager.memberEpoch()) {
+
+                final GroupMetadataUpdateEvent currentGroupMetadataUpdateEvent = new GroupMetadataUpdateEvent(
+                    membershipManager.memberEpoch(),
+                    previousGroupMetadataUdateEvent == null ? membershipManager.memberId() : previousGroupMetadataUdateEvent.memberId()
+                );
+                this.backgroundEventHandler.add(currentGroupMetadataUpdateEvent);
+                previousGroupMetadataUdateEvent = currentGroupMetadataUpdateEvent;
+            }
```

Comment:
@dajac @lucasbru @kirktrue @AndrewJSchofield Could you please sanity check this code snippet?
The heartbeat request manager should only send an group metadata update event if it is the first heartbeat response it gets (i.e. `previousGroupMetadataUdateEvent == null`), because that sets the member ID and the member ID never changes after this. Or the member epoch changed in the last heartbeat response (i.e., previousGroupMetadataUdateEvent.memberEpoch() != membershipManager.memberEpoch()).

Does this make sense?

Reply:
Can't the `memberId` change when I rejoin the group?

Reply:
I think you are right! Let me change that.